src/sentry/services/hybrid_cloud/import_export/impl.py

```diff
@@ -0,0 +1,372 @@
+            possible_relocation_scopes = model.get_possible_relocation_scopes()
+            includable = possible_relocation_scopes & allowed_relocation_scopes
+            if not includable or model._meta.proxy:
+                return RpcExportError(
+                    kind=RpcExportErrorKind.UnexportableModel,
+                    on=InstanceID(model_name),
+                    reason=f"The model `{str(batch_model_name)}` is not exportable",
+                )
+
+            max_pk = from_pk
+            out_pk_map = PrimaryKeyMap()
+            filters: List[Filter] = []
+            for fb in filter_by:
+                if NormalizedModelName(fb.model_name) == batch_model_name:
+                    filters.append(fb.from_rpc())
+
+            def filter_objects(queryset_iterator):
+                # Intercept each value from the queryset iterator, ensure that it has the correct
+                # relocation scope and that all of its dependencies have already been exported. If
+                # they have, store it in the `pk_map`, and then yield it again. If they have not, we
+                # know that some upstream model was filtered out, so we ignore this one as well.
+                for item in queryset_iterator:
+                    if not item.get_relocation_scope() in allowed_relocation_scopes:
+                        continue
+
+                    model = type(item)
+                    model_name = get_model_name(model)
+
+                    # Make sure this model is not explicitly being filtered.
+                    for f in filters:
+                        if f.model == model and getattr(item, f.field, None) not in f.values:
+                            break
+                    else:
+                        # Now make sure its not transitively filtered either.
+                        for field, foreign_field in deps[model_name].foreign_keys.items():
+                            dependency_model_name = get_model_name(foreign_field.model)
+                            field_id = field if field.endswith("_id") else f"{field}_id"
+                            fk = getattr(item, field_id, None)
+                            if fk is None:
+                                # Null deps are allowed.
+                                continue
+                            if in_pk_map.get_pk(dependency_model_name, fk) is None:
+                                # The foreign key value exists, but not found! An upstream model
+                                # must have been filtered out, so we can filter this one out as
+                                # well.
+                                break
+                        else:
+                            # For models that may have circular references to themselves (unlikely),
+                            # keep track of the new pk in the input map as well.
+                            nonlocal max_pk
+                            max_pk = item.pk
```

Comment:
At the moment, yes (see: `queryset = model._base_manager.filter(q).order_by(pk_name)` in `yield_objects()`). However, I concede that this may not always be true (I've considered ordering by other globally unique fields like email or slug where possible), so I've added a check here to ensure we return the true max even if this changes.